File: week1/week1_practice.py
"""

First week's practice for Google STEP:
与えられてあ16文字からなるべく長い単語を出力するプログラムをつくろう！
Given 16 letters and try to find all words which len(word) > 8 that output from the combination of this 16 words.
Set the length limit in order to decrease the number of output

Example:
Input: aehmnooqqrrstzzz
Output:
astronomer
hoarstone
anorthose
resonator

"""


def create_lst(path):
    letter = []
    f = open(path, "r")
    for line in f.readlines():
        letter.append(line.strip())
    return letter


def counts(string):
    d = {}
    for c in string:
        d[c] = d.get(c, 0) + 1
    return d


def valid(input_str, word):
    input_lst = counts(input_str)
    word_lst = counts(word)
    for key in word_lst:
        if key not in input_lst:
            return False
        if word_lst[key] > input_lst[key]:
            return False
    return True


def main():
    input_str = input("please input a word:")
    letters = create_lst("/usr/share/dict/words")
    for word in letters:
        if valid(input_str, word):
            if len(word) > 8:
                print(word)


if __name__ == "__main__":
    main()

File: week1/week1_practice.cpp


// First week's practice for Google STEP:
// Given 16 letters and try to find all words which len(word) > 8 that output from 16 letters.
// Set the length limit in order to decrease the number of output
//
// Example:
// Input: aehmnooqqrrstzzz
// Output:
// astronomer
// hoarstone
// anorthose
// resonator


#include <iostream>
#include <map>
#include <string>
#include <sstream>
#include <fstream>
#include <vector>
using namespace std;
map<char,int> count(string s);


vector<string> create_lst(string path){
    vector<string> dict;
    string word;
    ifstream file(path);
    if (file.is_open()){
        while (getline(file,word)){
            dict.push_back(word);
        }
    }
    return dict;
}


map<char,int> count(string s){
    map<char, int> d;
    for (int i = 0; i < s.size(); i++) {
        if(d.count(s[i]) == 0){
            d[s[i]] = 1;
        }
        else d[s[i]]++;
    }
    return d;
}


bool valid(string input, string word){
    map<char, int> input_dict = count(input);
    map<char, int> word_dict = count(word);
    for(map<char, int>::iterator iter = word_dict.begin(); iter != word_dict.end();iter++){
        if (!input_dict.count(iter->first)) {
            return 0;
        }
        if (word_dict[iter->first] > input_dict[iter->first]) {
            return 0;
        }
    }
    return true;
}


int main() {
    vector<string>dict = create_lst("/usr/share/dict/words");
    string input;
    cout <<"please input: ";
    cin >> input;
    for (vector<string>::iterator iter=dict.begin(); iter != dict.end(); iter++) {
        if (valid(input, *iter)) {
            if ((*iter).length() > 8) {
                cout << (*iter) <<endl;
            }
        }
    }
}
